Source/Provers/SMTLib/SMTLibProcess.cs

```diff
@@ -427,15 +428,7 @@ async IAsyncEnumerable<SExpr> ParseSExprs(bool top)
 
     Task<string> ReadProver()
     {
-      lock (this) {
-        if (proverOutput.TryDequeue(out var result)) {
-          return Task.FromResult(result);
-        }
-
-        var taskCompletionSource = new TaskCompletionSource<string>();
-        outputReceivers.Enqueue(taskCompletionSource);
-        return taskCompletionSource.Task;
-      }
+      return proverOutput.Dequeue(CancellationToken.None);
```

Comment:
I'm not using one on purpose. I've added a comment to explain why.

> This might result in a leak of tasks.

What do you mean by that?

Reply:
I might be wrong, but "Dequeue" enqueues a customer on a list of consumers, and this consumer consists of roughly the continuation of what happens after `ReadProver`
Perhaps should should also set the consumers to cancel when the items are cleared?

Reply:
Your answer is detailed but I am not sure it answers my concern.
I'm probably misunderstanding a point about tasks, I thought that if a task is not canceled or raise an exception, it's going to remain in memory forever. But in our case, since we call CheckerDied which eventually kills the process, even though the task is not cancelled, it will still raise an exception.

Reply:
> I thought that if a task is not canceled or raise an exception, it's going to remain in memory forever.

I'm not sure I understand what you're unsure of, but I'll share some related thoughts.

My understanding is that a Task is stored on the call stack of whatever thread it's being executed on, and when the Task completes its `ExecuteEntry` method then the call frame that references the Task is popped and the Task can be GC'd.

If there are continuations attached to the Task, then those will execute after the Task has reached some post running state. If the continuations are executed on a different thread than the Task then I imagine the Task will already get GC'd while the continuations run.

> But in our case, since we call CheckerDied which eventually kills the process, even though the task is not cancelled, it will still raise an exception.

In the case of `SMTLibProcess`, the Tasks are referenced by the `Queue<TaskCompletionSource<string>> outputReceivers` field stored in `SMTLibProcess`. When a `Checker` gets GC'd, so will its `SMTLibProcess` and the `SMTLibProcess.prover` field, and then the Task references in `outputReceivers` are lost. The Tasks don't complete with an Exception but their continuations are GC'd and never executed.